File: server/controllers/tasks.js
import express from 'express';
import { Task } from '../models/Task.js';
import { requireAuth } from '../middleware/auth.js';

const router = express.Router();

// Get all tasks for a project
router.get('/projects/:projectId/tasks', requireAuth, async (req, res) => {
  try {
    const tasks = await Task.find({ projectId: req.params.projectId });
    res.json(tasks);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Create task
router.post('/projects/:projectId/tasks', requireAuth, async (req, res) => {
  try {
    const { title, description, status, progress, priority, assignee, responsable, type, startDate, endDate } = req.body;
    console.log('[tasks] POST /projects/' + req.params.projectId + ' body:', JSON.stringify(req.body));
    const task = await Task.create({
      projectId: req.params.projectId,
      title,
      description,
      status: status || 'Non démarré',
      progress: typeof progress === 'number' ? progress : (progress ? Number(progress) : 0),
      priority: priority || 'Normal',
      type: type || 'Normal',
      assignee,
      responsable: responsable || '',
      startDate,
      endDate
    });
    console.log('[tasks] Created task id:', task._id, 'type/status:', task.type, task.status);
    res.status(201).json(task);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Update task
router.put('/tasks/:taskId', requireAuth, async (req, res) => {
  try {
    const { title, description, status, progress, priority, assignee, responsable, type, startDate, endDate } = req.body;
    console.log('[tasks] PUT /tasks/' + req.params.taskId + ' body:', JSON.stringify(req.body));
    const task = await Task.findByIdAndUpdate(
      req.params.taskId,
      { title, description, status, progress, priority, assignee, responsable, type, startDate, endDate },
      { new: true }
    );
    console.log('[tasks] Updated task id:', task?._id, 'type/status:', task?.type, task?.status);
    res.json(task);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Delete task
router.delete('/tasks/:taskId', requireAuth, async (req, res) => {
  try {
    await Task.findByIdAndDelete(req.params.taskId);
    res.json({ ok: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: server/controllers/tasks.py
from flask import Blueprint, request, jsonify
from pymongo import MongoClient
from bson.objectid import ObjectId
from middleware.auth import require_auth

bp = Blueprint("tasks", __name__)
client: MongoClient = None

def init(mongo_client):
    global client
    client = mongo_client


def oid(v):
    return str(v) if isinstance(v, ObjectId) else v


def sanitize(doc):
    return {k: oid(v) if isinstance(v, ObjectId) else v for k, v in (doc or {}).items()}


@bp.get("/api/projects/<pid>/tasks")
@require_auth
def list_tasks(pid):
    db = client.get_database()
    tasks = [sanitize(t) for t in db.tasks.find({"projectId": pid})]
    return jsonify(tasks)


@bp.post("/api/projects/<pid>/tasks")
@require_auth
def add_task(pid):
    db = client.get_database()
    d = request.json or {}
    t = {
        "projectId": pid,
        "title": d.get("title"),
        "description": d.get("description"),
        "status": d.get("status", "Non démarré"),
        "progress": d.get("progress", 0),
        "priority": d.get("priority", "Normal"),
        "type": d.get("type", "Normal"),
        "assignee": d.get("assignee"),
        "responsable": d.get("responsable"),
        "startDate": d.get("startDate"),
        "endDate": d.get("endDate"),
    }
    res = db.tasks.insert_one(t)
    task = db.tasks.find_one({"_id": res.inserted_id})
    return jsonify(sanitize(task)), 201


@bp.put("/api/tasks/<tid>")
@require_auth
def update_task(tid):
    db = client.get_database()
    d = request.json or {}
    db.tasks.update_one({"_id": ObjectId(tid)}, {"$set": {
        "title": d.get("title"),
        "description": d.get("description"),
        "status": d.get("status"),
        "progress": d.get("progress"),
        "priority": d.get("priority"),
        "type": d.get("type"),
        "assignee": d.get("assignee"),
        "responsable": d.get("responsable"),
        "startDate": d.get("startDate"),
        "endDate": d.get("endDate"),
    }})
    t = db.tasks.find_one({"_id": ObjectId(tid)})
    return jsonify(sanitize(t))


@bp.delete("/api/tasks/<tid>")
@require_auth
def delete_task(tid):
    db = client.get_database()
    db.tasks.delete_one({"_id": ObjectId(tid)})
    return jsonify({"ok": True})